README.md
[![Published on NPM](https://img.shields.io/npm/v/@advanced-rest-client/error-message.svg)](https://www.npmjs.com/package/@advanced-rest-client/error-message)

[![Build Status](https://travis-ci.org/advanced-rest-client/error-message.svg?branch=stage)](https://travis-ci.org/advanced-rest-client/error-message)

[![Published on webcomponents.org](https://img.shields.io/badge/webcomponents.org-published-blue.svg)](https://www.webcomponents.org/element/@advanced-rest-client/error-message)

# error-message;

Renders standarized error message with icon.

```html
<error-message iconprefix="arc" icon="sentiment-neutral">
  <p>This is neutral information</p>
</error-message>
```

### API components

This components is a part of [API components ecosystem](https://elements.advancedrestclient.com/)

## Usage

### Installation
```
npm install --save @advanced-rest-client/error-message
```

### In an html file

```html
<html>
  <head>
    <script type="module">
      import '@advanced-rest-client/error-message/error-message.js';
    </script>
  </head>
  <body>
    <error-message></error-message>
  </body>
</html>
```

### In a LitElement template

```js
import { LitElement, html } from 'lit-element';
import '@advanced-rest-client/error-message/error-message.js';

class SampleElement extends LitElement {
  render() {
    return html`<error-message></error-message>`;
  }
}
customElements.define('sample-element', SampleElement);
```

### In a Polymer 3 template

```js
import { PolymerElement, html } from '@polymer/polymer';
import '@advanced-rest-client/error-message/error-message.js';

class SampleElement extends PolymerElement {
  static get template() {
    return html`
    <error-message></error-message>
    `;
  }
}
customElements.define('sample-element', SampleElement);
```

## Development

```sh
git clone https://github.com/advanced-rest-client/error-message
cd error-message
npm install
```

### Running the demo locally

```sh
npm start
```

### Running the tests
```sh
npm test
```


CHANGELOG.md
<a name="0.1.1"></a>
## 0.1.1 (2017-01-03)




# [3.0.0](https://github.com/advanced-rest-client/error-message/compare/0.1.1...3.0.0) (2019-07-19)


### Breaking

* Updating styles to remove deprecated @apply function ([14f6c6f15061c9a93620db38c3914835e3134cb6](https://github.com/advanced-rest-client/error-message/commit/14f6c6f15061c9a93620db38c3914835e3134cb6))
* Upgrading to Polymer 3 ([ed727072295dbebe4a2afb17db82ac819f0b882b](https://github.com/advanced-rest-client/error-message/commit/ed727072295dbebe4a2afb17db82ac819f0b882b))

### build

* bumping version ([92517f13b10ae2987598524be09a2b959d9e3d1c](https://github.com/advanced-rest-client/error-message/commit/92517f13b10ae2987598524be09a2b959d9e3d1c))

### Docs

* Updating docs ([80cacb207dc4839836c904fd519c5b00a84fbba0](https://github.com/advanced-rest-client/error-message/commit/80cacb207dc4839836c904fd519c5b00a84fbba0))

### refactor

* dropping support for Polymer and replacing with LitElement ([582e7818a0aa2ada8e7f9d96689d23f4b3610468](https://github.com/advanced-rest-client/error-message/commit/582e7818a0aa2ada8e7f9d96689d23f4b3610468))

### Update

* Changing import paths to match web spec ([544ec4c2c565587848e624533eb29283ec40ff1d](https://github.com/advanced-rest-client/error-message/commit/544ec4c2c565587848e624533eb29283ec40ff1d))
* Updated Travis configuration to connect to Sauce Labs. ([2e609233d63f13b816911ad99d8826a991dc7c5e](https://github.com/advanced-rest-client/error-message/commit/2e609233d63f13b816911ad99d8826a991dc7c5e))



## [3.0.1](https://github.com/advanced-rest-client/error-message/compare/0.1.1...3.0.1) (2019-08-31)


### Bug Fixes

* fixing linter error and missing depdnency ([e04d05f](https://github.com/advanced-rest-client/error-message/commit/e04d05f))





test/index.d.ts
/**
 * DO NOT EDIT
 *
 * This file was automatically generated by
 *   https://github.com/Polymer/tools/tree/master/packages/gen-typescript-declarations
 *
 * To modify these typings, edit the source file(s):
 *   test/index.html
 */


// tslint:disable:variable-name Describing an API that's defined elsewhere.



test/basic-test.d.ts
/**
 * DO NOT EDIT
 *
 * This file was automatically generated by
 *   https://github.com/Polymer/tools/tree/master/packages/gen-typescript-declarations
 *
 * To modify these typings, edit the source file(s):
 *   test/basic-test.html
 */


// tslint:disable:variable-name Describing an API that's defined elsewhere.



test/error-message.test.js
import { fixture, assert } from '@open-wc/testing';
import '../error-message.js';

const style = document.createElement('style');
style.innerText = `.styled {
  --error-message-color: rgb(10, 10, 10);
  --error-message-icon-color: rgb(20, 20, 20);
}`;
document.head.appendChild(style);

describe('<error-message>', function() {
  async function basicFixture() {
    return (await fixture(`<error-message>
      <p id="message">Test message</p>
    </error-message>`));
  }

  async function styledFixture() {
    return (await fixture(`<error-message class="styled">
      <p id="message">Test message</p>
    </error-message>`));
  }

  describe('iconValue', () => {
    let element;
    beforeEach(async () => {
      element = await basicFixture();
    });

    it('Has default value', () => {
      const result = element.iconValue;
      assert.equal(result, 'arc:sentiment-very-dissatisfied');
    });

    it('Changes the prefix', () => {
      element.iconPrefix = 'other';
      const result = element.iconValue;
      assert.equal(result, 'other:sentiment-very-dissatisfied');
    });

    it('Ignores prefix when not set', () => {
      element.iconPrefix = '';
      const result = element.iconValue;
      assert.equal(result, 'sentiment-very-dissatisfied');
    });

    it('Changes the icon', () => {
      element.icon = 'other';
      const result = element.iconValue;
      assert.equal(result, 'arc:other');
    });
  });

  describe('Styling', () => {
    let element;
    beforeEach(async () => {
      element = await styledFixture();
    });

    it('Styles the message', () => {
      const color = getComputedStyle(element.children[0]).color;
      assert.equal(color, 'rgb(10, 10, 10)');
    });

    it('Styles the message', () => {
      const icon = element.shadowRoot.querySelector('.error-icon');
      const color = getComputedStyle(icon).color;
      assert.equal(color, 'rgb(20, 20, 20)');
    });
  });

  describe('a11y', () => {
    it('is accessible', async () => {
      const element = await basicFixture();
      await assert.isAccessible(element);
    });
  });
});


error-message.d.ts
/**
 * DO NOT EDIT
 *
 * This file was automatically generated by
 *   https://github.com/Polymer/tools/tree/master/packages/gen-typescript-declarations
 *
 * To modify these typings, edit the source file(s):
 *   error-message.js
 */


// tslint:disable:variable-name Describing an API that's defined elsewhere.
// tslint:disable:no-any describes the API as best we are able today

import {LitElement, html, css} from 'lit-element';

declare namespace UiElements {

  /**
   * `<error-message>` A standarized error information
   *
   * ### Example
   * ```
   * <error-message icon="warning" iconprefix="myset">
   *   <p>This is a warning!</p>
   * </error-message>
   * ```
   *
   * ### Styling
   * `<error-message>` provides the following custom properties and mixins for styling:
   *
   * Custom property | Description | Default
   * ----------------|-------------|----------
   * `--error-message-icon-color` | Fill color of the icon. Only if an icon is a SVG icon. | `rgba(0, 0, 0, 0.56)`
   * `--error-message-color` | Color of the message text. | `--google-red-500`
   * `--error-message-background-color` | Background color of the component | ``
   * `--arc-font-subhead-font-size` | Font size of the message (theme item) | ``
   * `--arc-font-subhead-font-weight` | Font weigth of the message (theme item) | ``
   * `--arc-font-subhead-line-height` | Font line height of the message (theme item) | ``
   */
  class ErrorMessage extends LitElement {
    readonly iconValue: String|null;

    /**
     * An icon to render. Do not use prefix here. Instead use the `iconPrefix`
     * property.
     *
     * Detaults to `sentiment-very-dissatisfied` from ARC icons set.
     */
    icon: string|null|undefined;

    /**
     * Icon prefix from the svg icon set. This can be used to replace the set
     * without changing the icon.
     *
     * Defaults to `arc`.
     */
    iconPrefix: string|null|undefined;
    constructor();
    render(): any;
  }
}

declare global {

  interface HTMLElementTagNameMap {
    "error-message": UiElements.ErrorMessage;
  }
}


karma.sl.config.js
/* eslint-disable import/no-extraneous-dependencies */
const merge = require('deepmerge');
const slSettings = require('@advanced-rest-client/testing-karma-sl/sl-settings.js');
const createBaseConfig = require('./karma.conf.js');

module.exports = (config) => {
  const cnf = slSettings();
  cnf.sauceLabs.testName = 'error-message';
  cnf.browsers = [
    'SL_Chrome',
    'SL_Chrome-1',
    'SL_Firefox',
    'SL_Firefox-1',
    'SL_Safari',
    'SL_EDGE'
  ];
  if (process.env.TRAVIS) {
    const buildLabel = 'TRAVIS #' + process.env.TRAVIS_BUILD_NUMBER + ' (' + process.env.TRAVIS_BUILD_ID + ')';

    cnf.browserStack = {
      build: buildLabel,
      tunnelIdentifier: process.env.TRAVIS_JOB_NUMBER
    };

    cnf.sauceLabs.build = buildLabel;
    cnf.sauceLabs.tunnelIdentifier = process.env.TRAVIS_JOB_NUMBER;
  }

  config.set(merge(createBaseConfig(config), cnf));

  return config;
};


error-message.js
/**
@license
Copyright 2016 The Advanced REST client authors <arc@mulesoft.com>
Licensed under the Apache License, Version 2.0 (the "License"); you may not
use this file except in compliance with the License. You may obtain a copy of
the License at
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
License for the specific language governing permissions and limitations under
the License.
*/
import { LitElement, html, css } from 'lit-element';
import '@polymer/iron-icon/iron-icon.js';
import '@advanced-rest-client/arc-icons/arc-icons.js';
/**
`<error-message>` A standarized error information

### Example
```
<error-message icon="warning" iconprefix="myset">
  <p>This is a warning!</p>
</error-message>
```

### Styling
`<error-message>` provides the following custom properties and mixins for styling:

Custom property | Description | Default
----------------|-------------|----------
`--error-message-icon-color` | Fill color of the icon. Only if an icon is a SVG icon. | `rgba(0, 0, 0, 0.56)`
`--error-message-color` | Color of the message text. | `--google-red-500`
`--error-message-background-color` | Background color of the component | ``
`--arc-font-subhead-font-size` | Font size of the message (theme item) | ``
`--arc-font-subhead-font-weight` | Font weigth of the message (theme item) | ``
`--arc-font-subhead-line-height` | Font line height of the message (theme item) | ``

@memberof UiElements
@element error-message
@customElement
@demo demo/index.html
*/
class ErrorMessage extends LitElement {
  static get styles() {
    return css`:host {
      display: flex;
      flex-direction: row;
      align-items: center;
      background-color: var(--error-message-background-color);
    }

    .error-icon {
      width: 128px;
      height: 128px;
      color: var(--error-message-icon-color, rgba(0, 0, 0, 0.56));
    }

    .error-desc ::slotted(*) {
      font-size: var(--arc-font-subhead-font-size);
      font-weight: var(--arc-font-subhead-font-weight);
      line-height: var(--arc-font-subhead-line-height);

      color: var(--error-message-color, var(--google-red-500));
    }`;
  }
  render() {
    const icon = this.iconValue;
    return html`<div>
      <iron-icon class="error-icon" .icon="${icon}"></iron-icon>
    </div>
    <div class="error-desc">
      <slot></slot>
    </div>`;
  }

  static get properties() {
    return {
      /**
       * An icon to render. Do not use prefix here. Instead use the `iconPrefix`
       * property.
       *
       * Detaults to `sentiment-very-dissatisfied` from ARC icons set.
       */
      icon: { type: String },
      /**
       * Icon prefix from the svg icon set. This can be used to replace the set
       * without changing the icon.
       *
       * Defaults to `arc`.
       */
      iconPrefix: { type: String }
    };
  }
  /**
   * @return {String} Computed value if the icon to be passed to `iron-icon`.
   * By default it is `arc:sentiment-very-dissatisfied`
   */
  get iconValue() {
    let icon = '';
    if (this.iconPrefix) {
      icon = this.iconPrefix + ':';
    }
    return icon + this.icon;
  }

  constructor() {
    super();
    this.icon = 'sentiment-very-dissatisfied';
    this.iconPrefix = 'arc';
  }
}
window.customElements.define('error-message', ErrorMessage);
